### ISSUE TITLE
unipept command: check correctness of -e option

### ISSUE THREAD

**Issue by pdawyndt** [unipept/unipept#265]
In processing metagenomics data sets, I would systematically avoid the use of the `-e` option, since we start from DNA reads (not peptides identified by mass spec). However, it seems that generally the use of the `-e` option yields far more protein hits for a given peptide than the same search without that option `-e` option. It is logical that the use of the `-e` option produces more hits than without, but my gut feeling is that now the number of differences is way too high.

Example:

```
$ fasta2prot < patient_TS19_CDS | sed -ne '10{p;q}' | prot2pept | peptfilter
YFGHIFSDEDK
NLLETGNMGR
VVELK
NGEYIPSFISIDK
LTNEVVAMK
AENAFIPR
GVELTEQLXTR
NEGIGK
AHEGI
$ fasta2prot < patient_TS19_CDS | sed -ne '10{p;q}' | prot2pept | peptfilter | unipept pept2lca
sequence,taxon_id,taxon_name,taxon_parent_id,taxon_rank
VVELK,1,root,1,no rank
NLLETGNMGR,976,Bacteroidetes,68336,phylum
LTNEVVAMK,171549,Bacteroidales,200643,order
$ fasta2prot < patient_TS19_CDS | sed -ne '10{p;q}' | prot2pept | peptfilter | unipept pept2lca -e
sequence,taxon_id,taxon_name,taxon_parent_id,taxon_rank
NEGLGK,1,root,1,no rank
VVELK,1,root,1,no rank
AENAFLPR,817,Bacteroides fragilis,816,species
NGEYLPSFLSLDK,817,Bacteroides fragilis,816,species
NLLETGNMGR,976,Bacteroidetes,68336,phylum
LTNEVVAMK,171549,Bacteroidales,200643,order
```

The problem seems real if we further explore the peptides in the above data set.

```
$ echo AENAFIPR | unipept pept2prot


$ echo AENAFIPR | unipept pept2prot -e
sequence,uniprot_id
AENAFLPR,I9B2A8
AENAFLPR,K1TAK5
AENAFLPR,K1SIZ1
AENAFLPR,I9BGY7
```

If I [BLAST](http://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastp&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome) this peptide, I find at least the following matches that have 100% coverage and 100% identity:
- [WP_021750589](http://www.ncbi.nlm.nih.gov/protein/545639344?report=genbank&log$=protalign&blast_rank=1&RID=P9AFYTPR01R) (Oscillibacter)
- [WP_005801812](http://www.ncbi.nlm.nih.gov/protein/492298822?report=genbank&log$=protalign&blast_rank=2&RID=P9AFYTPR01R) (Bacteroides fragilis)

These hits correspond to the following four hits in Uniprot
- [U2R6X9](http://www.uniprot.org/uniprot/U2R6X9) (Oscillibacter sp.)
- [U2RF28](http://www.uniprot.org/uniprot/U2RF28) (Oscillibacter sp.)
- [I9B2A8](http://www.uniprot.org/uniprot/I9B2A8) (Bacteroides fragilis)
- [I9BGY7](http://www.uniprot.org/uniprot/I9BGY7) (Bacteroides fragilis)

The latter two are clearly also found if the `-e` option is used, but since they match the original version of the peptide, they should also have been found without the `-e` option. The former two records have been added to GenBank in September 2013, and have probably been added to Uniprot in November 2013. As such, it might well be that they are not available yet in the current version of the Unipept database.

Check in web interface of Unipept:
- 4 hits are retrieved in case I and L are not equated (same four hits as above in case the `-e` option is used)
- same 4 hits are retrieved in case I and L are equated

I double checked, and all 4 protein records do have an exact match with the peptide sequence `AENAFIPR`

**Conclusion: current implementation of the `-e` option in the command line tools is not correct**

_[Original issue](https://github.ugent.be/unipept/unipept/issues/265) by @pdawyndt on Fri May 02 2014 at 21:01._
_Closed by an unknown user on Sat May 10 2014 at 13:55._

**Comment by pdawyndt** [comment]
Here's an overview of the lowest common ancestors (LCA) derived from the peptides of the protein analyzed above. Peptides that were not matched in Unipept have been checked further by BLAST:

```
YFGHIFSDEDK         -> (1)
NLLETGNMGR          -> Bacteroidetes (phylum)
VVELK               -> root
NGEYIPSFISIDK       -> Bacteroides fragilis (species)
LTNEVVAMK           -> Bacteroidales (order)
AENAFIPR            -> Bacteroides fragilis (species)
GVELTEQLXTR         -> no exact matches found by BLAST
NEGIGK              -> root
AHEGI               -> (2)
```
1. lots of BLAST hits -- most _Bacteroides_ species -- but probably no tryptic peptide due to fact that start of protein was not in read; can be found in Unipept for example as the last part of the peptide `QKPDLDRPYFGHLFSDEDK`
2. lots of BLAST hits, but no tryptic peptides due to fact that end of protein was not in read; we can not filter this peptide out, since trypsine breaks after `K` or `R`, so it might well be that this part was a tryptic peptide


_[Original comment](https://github.ugent.be/unipept/unipept/issues/265#issuecomment-9140) by @pdawyndt on Sat May 03 2014 at 11:35._

**Comment by pdawyndt** [comment]
@bmesuere : can you lookup some examples where equating I and L finds more matches than without (for example by looking up some peptides that have a different LCA in both cases)


_[Original comment](https://github.ugent.be/unipept/unipept/issues/265#issuecomment-9245) by @pdawyndt on Tue May 06 2014 at 09:44._

**Comment by bmesuere** [comment]
```
mysql> SELECT `sequences`.`sequence`, uniprot_entries.uniprot_accession_number, uniprot_entries.taxon_id FROM `sequences` INNER JOIN `peptides` ON `peptides`.`sequence_id` = `sequences`.`id` INNER JOIN `uniprot_entries` ON `uniprot_entries`.`id` = `peptides`.`uniprot_entry_id` WHERE `sequences`.`sequence` IN ('AENAFIPR');
```

should be

```
mysql> SELECT `sequences`.`sequence`, uniprot_entries.uniprot_accession_number, uniprot_entries.taxon_id FROM `sequences` INNER JOIN `peptides` ON `peptides`.`original_sequence_id` = `sequences`.`id` INNER JOIN `uniprot_entries` ON `uniprot_entries`.`id` = `peptides`.`uniprot_entry_id` WHERE `sequences`.`sequence` IN ('AENAFIPR');
```

`sequence_id` -> `original_sequence_id`


_[Original comment](https://github.ugent.be/unipept/unipept/issues/265#issuecomment-9311) by @bmesuere on Wed May 07 2014 at 14:48._

**Comment by bmesuere** [comment]
```
mysql> select * from sequences where lca != lca_il limit 10;
+-----+---------------------+--------+--------+
| id  | sequence            | lca    | lca_il |
+-----+---------------------+--------+--------+
|  33 | LDSVLGK             |  10492 |      1 |
|  97 | YTQLYK              |      2 |      1 |
| 181 | LLSSLEQK            |  51511 |   2759 |
| 204 | ENFVYLAK            |   3760 |  35493 |
| 225 | DNLTLWTSDLPEDGEDSQK |   3760 |  71275 |
| 246 | EMQPTHPLR           |   9601 |   7776 |
| 294 | LAELTSQR            |    780 |      2 |
| 305 | FLAVGYVDDTQFVR      |  59463 |      1 |
| 321 | CVEWLR              |   7776 |   2759 |
| 345 | VGNLQLSR            | 147550 |      1 |
+-----+---------------------+--------+--------+
```


_[Original comment](https://github.ugent.be/unipept/unipept/issues/265#issuecomment-9312) by @bmesuere on Wed May 07 2014 at 14:51._

**Comment by pdawyndt** [comment]
`@tvwillem` The `-e` option now seems to be supported adequately for the `pept2prot` and `pept2taxa` subcommands, but it still produces the wrong answer if used with the `pept2lca` subcommand. Can you check what needs to be done to correct the `pept2lca` subcommand?

Here's a use case showing that the `-e` option works as it should for the `pept2prot` and `pept2taxa` subcommands, but not for the `pept2lca` subcommand.

```
$ echo YTQLYK | unipept pept2prot
sequence,uniprot_accession_number,taxon_id
YTQLYK,I7EN32,391619
YTQLYK,I7EQH0,383629
YTQLYK,Q5LQN3,246200
YTQLYK,B6BCU2,439496
YTQLYK,D8DUV5,752555
YTQLYK,A4EW05,388739
YTQLYK,B7QPA7,439497
YTQLYK,A3SQ71,89187
YTQLYK,E4TV67,643867
YTQLYK,A3XBE4,314262
$ echo YTQLYK | unipept pept2prot -e
sequence,uniprot_accession_number,taxon_id
YTQLYK,P0C9E9,561444
YTQLYK,P0C9F2,561443
YTQLYK,P0C9F1,10500
YTQLYK,I7EN32,391619
YTQLYK,I7EQH0,383629
YTQLYK,Q5LQN3,246200
YTQLYK,B6BCU2,439496
YTQLYK,D8DUV5,752555
YTQLYK,A4EW05,388739
YTQLYK,B7QPA7,439497
YTQLYK,E0WMD2,874269
YTQLYK,A3SQ71,89187
YTQLYK,E4TV67,643867
YTQLYK,A9JLJ2,443878
YTQLYK,A3XBE4,314262
$ echo YTQLYK | unipept pept2taxa
sequence,taxon_id,taxon_name,taxon_parent_id,taxon_rank
YTQLYK,89187,Roseovarius nubinhibens ISM,314263,no rank
YTQLYK,246200,Ruegeria pomeroyi DSS-3,89184,no rank
YTQLYK,314262,Roseobacter sp. MED193,2433,species
YTQLYK,383629,Phaeobacter gallaeciensis 2.10,60890,no rank
YTQLYK,388739,Roseobacter sp. SK209-2-6,2433,species
YTQLYK,391619,Phaeobacter gallaeciensis DSM 17395 = CIP 105210,60890,no rank
YTQLYK,439496,Rhodobacterales bacterium Y4I,285892,species
YTQLYK,439497,Ruegeria sp. R11,97050,species
YTQLYK,643867,Marivirga tractuosa DSM 4126,1006,no rank
YTQLYK,752555,Prevotella bryantii B14,77095,no rank
$ echo YTQLYK | unipept pept2taxa -e
sequence,taxon_id,taxon_name,taxon_parent_id,taxon_rank
YTQLYK,10500,African swine fever virus Malawi LIL 20/1,10497,no rank
YTQLYK,89187,Roseovarius nubinhibens ISM,314263,no rank
YTQLYK,246200,Ruegeria pomeroyi DSS-3,89184,no rank
YTQLYK,314262,Roseobacter sp. MED193,2433,species
YTQLYK,383629,Phaeobacter gallaeciensis 2.10,60890,no rank
YTQLYK,388739,Roseobacter sp. SK209-2-6,2433,species
YTQLYK,391619,Phaeobacter gallaeciensis DSM 17395 = CIP 105210,60890,no rank
YTQLYK,439496,Rhodobacterales bacterium Y4I,285892,species
YTQLYK,439497,Ruegeria sp. R11,97050,species
YTQLYK,443878,African swine fever virus OURT 88/3,10497,no rank
YTQLYK,561443,African swine fever virus tick/South Africa/Pretoriuskop Pr4/1996,10497,no rank
YTQLYK,561444,African swine fever virus warthog/Namibia/Wart80/1980,10497,no rank
YTQLYK,643867,Marivirga tractuosa DSM 4126,1006,no rank
YTQLYK,752555,Prevotella bryantii B14,77095,no rank
YTQLYK,874269,African swine fever virus Georgia 2007/1,10497,no rank
$ echo YTQLYK | unipept pept2lca
sequence,taxon_id,taxon_name,taxon_parent_id,taxon_rank
YTQLYK,1,root,1,no rank
$ echo YTQLYK | unipept pept2lca -e
sequence,taxon_id,taxon_name,taxon_parent_id,taxon_rank
YTQLYK,1,root,1,no rank
```

The `echo YTQLYK | unipept pept2lca` command should give the superkingdom _Bacteria_ as LCA. See the web interface for the proper outcome:
- [identification of YTQLYK with I and L considered to be differently](http://unipept.ugent.be/sequences/YTQLYK): Bacteria (superkingdom)
- [identification of YTQLYK with I and L considered to be equal](http://unipept.ugent.be/sequences/YTQLYK/equateIL): root (no rank)


_[Original comment](https://github.ugent.be/unipept/unipept/issues/265#issuecomment-9428) by @pdawyndt on Sat May 10 2014 at 10:40._

**Comment by bmesuere** [comment]
It should work for pept2taxa, I did however overlook something on pept2lca.


_[Original comment](https://github.ugent.be/unipept/unipept/issues/265#issuecomment-9429) by ghost on Sat May 10 2014 at 10:44._

**Comment by bmesuere** [comment]
pept2lca fixed in 110b2c2


_[Original comment](https://github.ugent.be/unipept/unipept/issues/265#issuecomment-9441) by ghost on Sat May 10 2014 at 13:55._
